### PR TITLE
BFD-2648: Introduce terraservice child module

### PR DESCRIPTION
**Pull request by mjburling**
**JIRA Ticket:**
[BFD-2648](https://jira.cms.gov/browse/BFD-2648)

**User Story or Bug Summary:**

As a BFD infrastructure engineer I want there to be a reusable module for the terraservices that handles basic variable definitions so that I can more easily work with the terraservices without having to repeat myself.

---

### What Does This PR Do?

Introduces a terraservice child module for reusability and modifies exist `main.tf` files in the `ops/terraform` directory to reference this module.

### What Should Reviewers Watch For?

If you're reviewing this PR, please check for these things in particular:

* Verify all PR security questions and checklists have been completed and addressed.

### What Security Implications Does This PR Have?

Submitters should complete the following questionnaire:

* If the answer to any of the questions below is **Yes**, then **you must** supply a link to the associated Security Impact Assessment (SIA), security checklist, or other similar document in Confluence here: **N/A**

    * Does this PR add any new software dependencies? 
      * [ ] Yes
      * [x] No
    * Does this PR modify or invalidate any of our security controls?
      * [ ] Yes
      * [x] No
    * Does this PR store or transmit data that was not stored or transmitted before?
      * [ ] Yes
      * [x] No

* If the answer to any of the questions below is **Yes**, then please add @StewGoin as a reviewer, and note that this PR **should not be merged** unless/until he also approves it.
    * Do you think this PR requires additional review of its security implications for other reasons?
      * [ ] Yes
      * [x] No

### What Needs to Be Merged and Deployed Before this PR?

This PR cannot be either merged or deployed until the following prerequisite changes have been fully deployed:

* N/A

### Submitter Checklist

I have gone through and verified that...:

* [x] I have named this PR and branch so they are [automatically linked](https://confluence.atlassian.com/adminjiracloud/integrating-with-development-tools-776636216.html) to the (most) relevant Jira issue. Ie: `BFD-123: Adds foo`
* [x] This PR is reasonably limited in scope, to help ensure that:
    1. It doesn't unnecessarily tie a bunch of disparate features, fixes, refactorings, etc. together.
    2. There isn't too much of a burden on reviewers.
    3. Any problems it causes have a small "blast radius".
    4. It'll be easier to rollback if that becomes necessary.
* [x] This PR includes any required documentation changes, including `README` updates and changelog / release notes entries.
* [x] The data dictionary has been updated with any field mapping changes, if any were made.
* [x] All new and modified code is appropriately commented, such that the what and why of its design would be reasonably clear to engineers, preferably ones unfamiliar with the project.
* [x] All tech debt and/or shortcomings introduced by this PR are detailed in `TODO` and/or `FIXME` comments, which include a JIRA ticket ID for any items that require urgent attention.
* [x] Reviews are requested from both:
    * At least two other engineers on this project, at least one of whom is a senior engineer or owns the relevant component(s) here.
    * Any relevant engineers on other projects (e.g. DC GEO, BB2, etc.).
* [x] Any deviations from the other policies in the [DASG Engineering Standards](https://github.com/CMSgov/cms-oeda-dasg/blob/master/policies/engineering_standards.md) are specifically called out in this PR, above.
    * Please review the standards every few months to ensure you're familiar with them.

### CODE REVIEW

**Comment by mjburling** [comment]
This **is** ready, but we'll be kinder to the would-be reviewers in the near term and pull it back into draft. We'll see this re-surface no later than two weeks from today.

File: ops/terraform/services/server/server-regression/main.tf [hunk 1]
```diff
@@ -1,20 +1,19 @@
...
+}
 
-  env              = terraform.workspace
-  established_envs = ["test", "prod-sbx", "prod"]
-  seed_env         = one([for x in local.established_envs : x if can(regex("${x}$$", local.env))])
-  is_ephemeral_env = !(contains(local.established_envs, local.env))
+locals {
+  default_tags = module.terraservice.default_tags
+  env          = module.terraservice.env
+  seed_env     = module.terraservice.seed_env
```
Comment (brianburton) [inline ops/terraform/services/server/server-regression/main.tf:17]: Was the ommission of `is_ephemeral_environment` deliberate?

File: ops/terraform/services/server/server-load/main.tf [hunk 2]
```diff
@@ -1,23 +1,22 @@
...
   account_id             = data.aws_caller_identity.current.account_id
   availability_zone_name = var.create_locust_instance ? data.aws_availability_zones.this.names[random_integer.this[0].result] : ""
 
-  env              = terraform.workspace
-  established_envs = ["test", "prod-sbx", "prod"]
-  seed_env         = one([for x in local.established_envs : x if can(regex("${x}$$", local.env))])
-  is_ephemeral_env = !(contains(local.established_envs, local.env))
+  default_tags = module.terraservice.default_tags
+  env          = module.terraservice.env
+  seed_env     = module.terraservice.seed_env
```
Comment (brianburton) [inline ops/terraform/services/server/server-load/main.tf:20]: Was the ommission of is_ephemeral_environment deliberate?

File: ops/terraform/services/_modules/bfd-terraservice/variables.tf [hunk 3]
```diff
@@ -0,0 +1,22 @@
+variable "environment_name" {
+  description = "The solution's environment name. Generally, `terraform.workspace`"
+  type        = string
+  validation {
+    // NOTE: validations can only refer to this scope, i.e. `var.environment_name` is valid but `local.established_envs`
```
Comment (lsmitchell) [inline ops/terraform/services/_modules/bfd-terraservice/variables.tf:5]:
```suggestion
    // NOTE: validations can only refer to this scope, i.e. `var.environment_name` is valid but `local.established_envs` is not
```
?

Comment (mjburling) [reply ops/terraform/services/_modules/bfd-terraservice/variables.tf]: Thanks!

File: ops/terraform/services/server/server-regression/main.tf (hunk 1 above)
Comment (mjburling) [reply ops/terraform/services/server/server-regression/main.tf:17]: That's intentional. This was mostly copy/pasta or _cargoculting_.

File: ops/terraform/services/server/server-load/main.tf (hunk 2 above)
Comment (mjburling) [reply ops/terraform/services/server/server-load/main.tf:20]: That's intentional. This was mostly copy/pasta or _cargoculting_.

**Review by malessi — APPROVED** [review]
Paired through `plan`ning these changes to all Terraservices and it all looked good.

**Review by mackec1 — APPROVED** [review]

**Pull request merged**